### ISSUE TITLE
Fully Zoom Out option while loading 

### ISSUE THREAD

**Issue by Arshad8464** [vasturiano/force-graph#94]
Hi 
I was trying to load the graph in a fully zoomed out position while first loading. But I couldn't get it through.
I tried the **zoom** function but I dunno how to calculate value for the fully zoomed out?
Is there any option to set the left, right, bottom and top position of the graph? If so I can get the node position to set these positions.

Can you please advise how to achieve it?

**Comment by vasturiano** [comment]
@Arshad8464 thanks for reaching out. 

There is no direct option to do this, but you can get there by reading in the current positions of your nodes and determine their bounding box. Then adjust the `zoom` according to the value that fits those coordinates within your canvas dimensions. You can even translate the graph sideways using `centerAt` if your bounding box is asymmetrical, for an optimal fit.

**Comment by Alexithemia** [comment]
I used a calculation to set a zoom value that is based off of how many nodes exist in the graph. With forces being used, the distance the nodes end up apart should be somewhat predictable, and you can set a base zoom level and have it increase when there are a certain amount of nodes, in steps or with an equation.

**Comment by tarun080698** [comment]
> I used a calculation to set a zoom value that is based off of how many nodes exist in the graph. With forces being used, the distance the nodes end up apart should be somewhat predictable, and you can set a base zoom level and have it increase when there are a certain amount of nodes, in steps or with an equation.

how did you calculate zoom based on the number of nodes in the graph?

**Comment by vasturiano** [comment]
@tarun080698 please note that there is now a built-in `zoomToFit` method that includes this functionality, so you don't need to custom code it.

**Comment by tarun080698** [comment]
Yes, I read the docs and that for highlighting the zoomToFit method again. But I used it as 
`graph.zoomToFit(2000, 100)`

the result is something like this... It is hiding and going out of the canvas view 
![CPT2106111748-1303x801](https://user-images.githubusercontent.com/53266806/121685168-3aeb1480-cadd-11eb-80f3-ac014ab28197.gif)
I want that my nodes should fit in the canvas when the graph is generated.

can you help me with that?

**Comment by vasturiano** [comment]
@tarun080698 it's possibly a timing issue. If you're requesting a `zoomToFit` before the graph has completely cooled down, it will move out of its boundaries by the time the animation is finished.
Have you looked into the example, for applying the fitting zoom **after** the graph motion has stopped?
https://vasturiano.github.io/force-graph/example/fit-to-canvas/

**Comment by tarun080698** [comment]
This works fine, it was the timing issue, you were right. 
![CPT2106130251-885x815](https://user-images.githubusercontent.com/53266806/121789305-88a57100-cbf2-11eb-9586-1cdc22e3dc8f.gif)

Now the problem is when I click on any node to focus on that node, somehow, it goes back to the **zoomToFit** position. Perhaps, I can probably think of it as when I click any node, the engine starts and stops again which causes the trigger. 
```
onNodeClick={(node, event) => {
    graph.centerAt(node.x, node.y, 1000)
    graph.zoom(6, 1000)
}}
```
It looks something like this
![CPT2106130257-882x801](https://user-images.githubusercontent.com/53266806/121789405-59433400-cbf3-11eb-9e31-a5f9a9a6a10f.gif)

How can I stop it? @vasturiano

**Comment by vasturiano** [comment]
Assuming you're applying the `zoomToFit` on `onEngineStop`, whenever you wish to cancel that functionality (which will happen everytime the engine re-heats and cools down again), you can simply re-assign the method. For example, whenever a node is click:
```js
myGraph.onNodeClick(() => {
  myGraph.onEngineStop(() => {});
  ...
})
```

Or, if you'd like to have the `zoomToFit` apply only once the first time around:
```js
myGraph.onEngineStop(() => {
  myGraph.zoomToFit(600);
  myGraph.onEngineStop(() => {});
});
```

**Comment by tarun080698** [comment]
Hey @vasturiano, I am actually using in it a ReactJs App. Using a reference to graph to call on the method like **zoom**, **zoomToFit** and others. Now when I try to call the **onEngineStop** using the same reference. i gives me error.

```
<ForceGraph2D
    ref={fgRef}
    graphData={props.signalData}
    height={props.heightEle}
    width={props.widthEle}
    nodeId="id"
    autoPauseRedraw={false}
    onNodeClick={(node, event) => {
        fgRef.current.centerAt(node.x, node.y, 1000)
        fgRef.current.zoom(6, 1000)
        fgRef.current.onEngineStop(() => {// do nothing })
     }}
     onEngineStop={() => {
        fgRef.current.zoomToFit(1000, 100)
        fgRef.current.onEngineStop = () => {// do nothing}}}
     cooldownTicks={100}
          />
```

If I write it like in the above snippet, it does not work. and if i try to write that as 
```
onEngineStop={() => {
    fgRef.current.zoomToFit(1000, 100)
    fgRef.current.onEngineStop(() => { // do nothing })
}}
```
it gives me the error **"onEngineStop is not a function."**

**Comment by vasturiano** [comment]
@tarun080698 that is expected, because `onEngineStop` is a component prop, not method.

You'd need to keep state of the first time it's run. Something like:
```js
const [engineStoppedOnce, setEngineStoppedOnce] = useState(false);
```
and then 
```js
onEngineStop={() => {
    if (!engineStoppedOnce) {
      fgRef.current.zoomToFit(1000, 100);
      setEngineStoppedOnce(true);
    }
}}
```

**Comment by tarun080698** [comment]
The error of **onEngineStop** seems to be resolved but my nodes are still spreading out of the canvas. 
![CPT2106140942-1303x801](https://user-images.githubusercontent.com/53266806/121838161-de624200-ccf4-11eb-91c2-e061422456bc.gif)

Another thing I want to know is that how can I specify the distance between nodes. like in the above reference, If I want to specify that two diff graphs entities should be nearer to each other.

**Comment by vasturiano** [comment]
@tarun080698 to prevent disconnected section to stray too far from each other, you can try setting the [maxDistance](https://github.com/d3/d3-force#manyBody_distanceMax) of the charge force. F.e.
```js
myGraph.d3Force('charge').maxDistance(80); // adjust to taste
```

As for the engine stop part, could you make a simple reproducible example on [codeSandbox](https://codesandbox.io/), otherwise it's impossible to debug.

**Comment by tarun080698** [comment]
@vasturiano, I have created an example with my data [here](https://codesandbox.io/s/nodes-going-out-of-canvas-issue-8zu6o). I found out that to limit zoom, I used **onZoom** which was causing this problem. Please verify my method in the provided sandbox. I want to control the zoom-in and zoom-out limit. So, in order to do that, I used **onZoom** and **zoom()** to achieve it.
If my method is wrong, please help me with that. Thanks

**Comment by tarun080698** [comment]
@vasturiano, yes I checked the new update. After updating the library, if i set **minZoom** in props and **zoomToFit()** in **onEnginerStop** then **zoomToFit** does not work and if i comment the **minZoom**,  **zoomToFit** works. 
I want to know if this is a limitation or something. I will write the code for the situation I explained. Check it [here](https://codesandbox.io/s/nodes-going-out-of-canvas-issue-8zu6o)

**Comment by vasturiano** [comment]
If you restrict the `minZoom` to `1` it means the framework will not be able to zoom out further than unity. That conflicts with the `zoomToFit` functionality that tries to zoom out to fit the graph but is blocked by this restriction. I would advise to relax the `minZoom` restriction to a value that can at least accommodate for your whole chart to fit in the view.

**Comment by tarun080698** [comment]
@vasturiano, that's exactly what I thought the reason for the **zoomToFit** not to work in some cases. I tried removing the min zoom and everything was working fine. Thanks for the assistance. Have a great time.

**Comment by tarun080698** [comment]
> @tarun080698 to prevent the disconnected section from stray too far from each other, you can try setting the [maxDistance](https://github.com/d3/d3-force#manyBody_distanceMax) of the charge force. F.e.
> 
> ```js
> myGraph.d3Force('charge').maxDistance(80); // adjust to taste
> ```
Hey, I tried this but it is setting my nodes closer not the different graph entities close.
![grp](https://user-images.githubusercontent.com/53266806/122223345-b3cce080-ced0-11eb-8cac-9630c197bed0.png)
As you can see in the image above, I want to reduce the distance marked by wide white lines.

**Comment by tarun080698** [comment]
@vasturiano I managed to achieve the above requirement. I want to inform you about the function name you suggested has changed from **maxDistance** to **distanceMax** I guess. You can confirm it here [here](https://github.com/d3/d3-force#manyBody_distanceMax). 

Updated code snippet: 
`myGraph.d3Force('charge').distanceMax(80); // adjust to taste`
Thanks for the idea and docs reference.

**Comment by vasturiano** [comment]
Ah yes, my bad. I misspelled the method name, it's always been `distanceMax`. Thanks for the correction. 👍 
I'll edit the comment above to not mislead other users.
